### PR TITLE
[path_provider] linux endorsement

### PR DESCRIPTION
**Pull request by TimWhiting**
## Description

This PR updates path_provider to endorse the dart Linux path provider manually, until automatic registration of dart plugins are working.

## Related Issues

flutter/flutter#41716

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?
Only to remove manual registration if anyone does that before this gets implemented.

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by TimWhiting** [comment]
@stuartmorgan 
FYI Here is the PR for the endorsement of the Linux path provider. 

Note that I need to manually register the instance, at least until flutter tool support for automatically generating a plugin registrant for dart only plugins happens. If that is in the plans? (I didn't see any conclusion on which approach you are planning on taking) [Design Document: Flutter tool support for federated plugins](https://docs.google.com/document/d/1G_K5vTMWD6CBD-0DDWhpAwkcCRKqmPH1wxlXu5FfRhk/edit#heading=h.pub7jnop54q0)

**Comment by stuartmorgan** [comment]
> Note that I need to manually register the instance, at least until flutter tool support for automatically generating a plugin registrant for dart only plugins happens. If that is in the plans?

Ideally yes, but it ended up having technical complexities, so we decided to go with this approach in the short-to-medium term to unblock being able to implement Dart-only plugins. https://github.com/flutter/flutter/issues/52267 is tracking a longer term solution.

**Comment by TimWhiting** [comment]
@stuartmorgan 
Ready for review

File: packages/path_provider/path_provider/test/path_provider_test.dart
Comment (stuartmorgan) [inline packages/path_provider/path_provider/test/path_provider_test.dart:29]: Please link to https://github.com/flutter/flutter/issues/52267 in any comment about workarounds so that it's easy for someone without full context to trace back to the relevant discussion and current state.

File: packages/path_provider/path_provider/lib/path_provider.dart
Comment (stuartmorgan) [inline packages/path_provider/path_provider/lib/path_provider.dart:29]: Do we need a `kIsWeb` check first? I'm not sure what `Platform.isLinux` returns when running Flutter for Web on Linux.

File: packages/path_provider/path_provider/lib/path_provider.dart
Comment (stuartmorgan) [inline packages/path_provider/path_provider/lib/path_provider.dart:24]: Per [my comment here](https://github.com/flutter/plugins/pull/2631#discussion_r406489237) I'm not sure if we want this hidden; there are legitimate reasons some could want to call this.

@tvolkert Who is the right person to weigh in on this at the moment with amirh OOO?

File: packages/path_provider/path_provider_linux/README.md
Comment (stuartmorgan) [inline packages/path_provider/path_provider_linux/README.md:14]: It seems odd to have the section be called "import the package" when that's not actually necessary.

I think this whole thing could just become:
```
## Usage

This package is already included as part of the `path_provider` package dependency, and will
be included when using `path_provider` as normal. You will need to use version 1.6.10 or newer.
```

File: packages/path_provider/path_provider_linux/example/test/widget_test.dart
Comment (stuartmorgan) [inline packages/path_provider/path_provider_linux/example/test/widget_test.dart:25]: Why was this necessary? Does this take more than 10 milliseconds to run?

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]

File: packages/path_provider/path_provider_linux/example/test/widget_test.dart
Comment (TimWhiting) [reply packages/path_provider/path_provider_linux/example/test/widget_test.dart:25]: The tests were being flaky, for some reason they were failing on CI. As there are no logical flaws in the test or the app code as far as I can tell I assumed the asynchronous code in the `initState` of `MyApp` that calls out to the plugin to get the various paths had not yet been run. Increasing the timeout seemed to solve the issue. Is there a better way to make sure all asynchronous events are finished being handled in a test?

File: packages/path_provider/path_provider/lib/path_provider.dart
Comment (TimWhiting) [reply packages/path_provider/path_provider/lib/path_provider.dart:33]: Renamed the platform override and made it user visible in my latest commit, I doubt that people will be using their own registered version of `path_provider` unless they are using a different embedding i.e. `go-flutter`, but it will mean that those implementations will have to add `disablePathProviderPlatformOverride = true` until we have automatic registration.

Comment (TimWhiting) [reply packages/path_provider/path_provider/lib/path_provider.dart:33]: @stuartmorgan 
I've made this user visible, and renamed / reformatted it some. Should we continue waiting for a reply? The only issue would be that removing it in the future would be technically a breaking change for anyone who is using it.

Comment (stuartmorgan) [reply packages/path_provider/path_provider/lib/path_provider.dart:33]: I was thinking about this last night, and was going to updating this morning actually. Since getting resolution on this is proving difficult, let's return it to being test-only. That means we're not adding to the public API surface, so it won't need the same level of discussion/approval. We can always add it later if it becomes necessary, without it blocking getting the initial implementation landed.

**Comment by TimWhiting** [comment]
@stuartmorgan 
Just made that setter visible for testing, and updated the dartdoc for it. Let me know if further adjustments should be made. I tried to not change the dartdoc too much, in case we want to go back to being user visible.

I think this should be ready to land, once all the tests pass.

**Review by stuartmorgan-g — APPROVED** [review]
LGTM!

**Comment by TimWhiting** [comment]
@stuartmorgan 
Do I need to merge in master with that latest merged pull request or can you merge this in without that?

**Comment by stuartmorgan** [comment]
There aren't an conflicts; no need to re-merge.

**Comment by stuartmorgan** [comment]
Once main CI has cycled green I should be able to publish the update.

**Comment by stuartmorgan-g** [comment]
path_provider 1.6.10 is live 🎉

**Comment by stuartmorgan-g** [comment]
I filed https://github.com/flutter/website/issues/8442 for updating the cookbook. Updating the path_provider docs to call out the necessary changes sounds good. Given that it's already published, I don't think there's much more we can do to mitigate the potential unintentional test breakage.

**Pull request merged**
